crates/ruff_graph/src/resolver.rs

```diff
@@ -0,0 +1,102 @@
+use crate::db::ModuleDb;
+use red_knot_python_semantic::{Module, ModuleName};
+use ruff_db::files::FilePath;
+
+use crate::collector::CollectedImport;
+use ruff_python_ast::name::{QualifiedName, QualifiedNameBuilder};
+use ruff_python_stdlib::identifiers::is_identifier;
+use std::borrow::Cow;
+
+/// Collect all imports for a given Python file.
+pub(crate) struct Resolver<'ast> {
+    module_path: Option<&'ast [String]>,
+    db: &'ast ModuleDb,
+}
+
+impl<'ast> Resolver<'ast> {
+    pub(crate) fn new(module_path: Option<&'ast [String]>, db: &'ast ModuleDb) -> Self {
+        Self { module_path, db }
+    }
+
+    pub(crate) fn resolve(&self, import: &'ast CollectedImport<'ast>) -> Option<&'ast FilePath> {
+        match import {
+            CollectedImport::Import(import) => Some(
+                resolve_module(self.db, import.segments())?
+                    .file()
+                    .path(self.db),
+            ),
+            CollectedImport::ImportFrom(import) => {
+                let module_path = if import.is_unresolved_import() {
+                    // Only fix is the module path is known, and the import doesn't extend beyond it.
+                    let module_path = from_relative_import(self.module_path?, import)?;
+
+                    // Require import to be a valid module:
+                    // https://python.org/dev/peps/pep-0008/#package-and-module-names
+                    if !module_path
+                        .segments()
+                        .iter()
+                        .all(|segment| is_identifier(segment))
```

Comment:
You can just reuse the logic we have in `ModuleName::new()` here as well, I think